models/User.js

```diff
@@ -59,44 +60,44 @@ UserSchema.methods.comparePassword = (candidatePassword, hashedPassword, done)=>
   });
 };
 
-
-
-UserSchema.pre('save', function(next) {
-  var _user = this;
-
-  // only hash the password if it has been modified (or is new)
-  if (!_user.isModified('password')){
-    return next();
-  }
-
-  var generateSalt = function() {
-    var d = Q.defer();
-    bcrypt.genSalt(SALT_WORK_FACTOR, function(err, salt) {
-      if(err){
-        return next(err);
+function generateSalt(rawPassword) {
+  return new Promise((resolve, reject)=> {
+    bcrypt.genSalt(SALT_WORK_FACTOR, (err, salt)=> {
+      if (err) {
+        return reject(err);
       }
-
-      d.resolve(salt);
+      resolve({ rawPassword, salt });
     });
-    return d.promise;
-  };
+  });
+}
 
-  var hashPassword = function(salt) {
-    var d = Q.defer();
-    bcrypt.hash(_user.password, salt, function(err, hash) {
-      if (err){
-        return next(err);
+function hashPassword({ rawPassword, salt }) {
+  return new Promise((resolve, reject)=> {
```

Comment:
__[jshint]__
*  'arrow function syntax (=>)' is only available in ES6 (use esnext option).